[D1] D1 RR Documentation (#20922)
cloudflare/cloudflare-docs: src/content/docs/d1/features/read-replication.mdx

---
title: Read replication
pcx_content_type: concept
sidebar:
  order: 2

---

import { GlossaryTooltip, Details } from "~/components"

D1 read replication is a feature which reduces the <GlossaryTooltip term = "request latency">request latency</GlossaryTooltip> for read requests for users who may be located far away from the <GlossaryTooltip term="primary database instance">primary database instance</GlossaryTooltip>.

You can use D1 read replication by using D1 Sessions API. A Session encapsulates all the queries from one logical session for your application. For example, a Session may correspond to all queries coming from a particular web browser session.

> By using Sessions API for read replication, all of your queries from a single Session read from a version of the database which is as up-to-date as your query. This ensures that the version of the database you are reading is logically consistent with your queries when using read replicas.

Reviewer: ```suggestion
By using Sessions API for read replication, all of your queries from a single Session read from a version of the database which is as up-to-date as your query requires. This ensures that the version of the database you are reading is logically consistent with your queries when using read replicas.
```

Author: Will fix manually along with other changes

The following Worker code uses D1 Sessions API to use read replicas, and allows you to test the decrease in query latency when using read replication.

> ```js collapse={8-15, 24-31}

Author: I only knew how to use JS, which is why I'd written it in JS 😅

This is a placeholder (like you've said), so I'll wait until I have more details about the "Deploy to Workers" button, and swap it in. Do you want the code snippet to be visible too? (Rather than just having the button)

> ```js collapse={8-15, 24-31}

Author: I've asked @lambrospetrou to move the code to `cloudflare` repo, as `GitHubCode` only sees inside `cloudflare` repo.

Pending `Deploy to Workers` button
Pending REST API - I'll need a D1 engineer to help with this 🙏🏻

	/run - Queries the table without using read replication
	/withsession - Queries the table using read replication (using "first-unconstrained")`
	  );
	}
};
```

{/* Deploy to Workers button */}

## Primary database instance vs read replicas

When using D1 without read replication, D1 routes all queries (both read and write) to a specific database instance in [one location in the world](/d1/configuration/data-location/), known as the primary database instance. The request latency is dependent on the physical closeness of a user to the primary database instance - it takes less time for light (information) to travel between the user and the primary database instance if that distance is shorter. Users located further away from this database instance experience the longest request latency.

When using read replication, D1 introduces multiple “almost up-to-date” copies of the primary database instance which only serve read requests, called <GlossaryTooltip term="read replica"> read replicas </GlossaryTooltip>. D1 creates the read replicas in multiple regions throughout the world [across the Cloudflare network](/d1/features/read-replication/#read-replica-locations).

> A user may be located far away from the primary database instance but close to a read replica. By sending their read requests to the read replica instead of the primary database instance, the user enjoys shorter read request response times. For example, the request latency when using the primary database instance may be 250 ms, versus 20 - 50 ms when using a read replica.

Author: Have removed the example (though, I think it would be good if we could provide a rough range of ms for improvements we expect them to see?)

Instead of combining to first para, I've rephrased to: `When D1 routes read requests to the read replica instead of the primary database instance, the user enjoys shorter read request response times.`

Please let me know if you think it's still misleading

D1 read replication offers [sequential consistency](https://jepsen.io/consistency/models/sequential). D1 creates a global order of all operations which have taken place on the database, and can identify the latest version of the database that a query has seen, using [bookmarks](/d1/reference/time-travel/#bookmarks). It then serves the query with a database instance that is at least as up-to-date as the query itself.

Sequential consistency has properties such as:

- **Monotonic reads**: If you perform two reads one after the other (read-1, then read-2), read-2 cannot read a version of the database prior to read-1.
- **Monotonic writes**: If you perform write-1 then write-2, all processes observe write-1 before write-2.
- **Writes follow reads**: If you read a value, then perform a write, the subsequent write must be based on the value that was just read.
- **Read my own writes**: If you write to the database, all subsequent reads will see the write.

> ## Use Sessions API

Reviewer: This section needs the most rework, some suggestions:

* think of this as a task to be done/example bank with copy/paste code snippets
* the tasks we should have code/explainer for:
** start session with no context (either `null` or `first-unconstrained` argument, explain why `first-unconstrained` is like a clean slate)
** start session with constraint
** start session from previous session context (where bookmarks fit in)

Reviewer: @Oxyjun few more tasks for you:
* check if replication is enabled
* disable replication

Author: Marking as resolved, as the structure has been implemented - will work off of the second round of review.

### Use read replication via Sessions API

By using <GlossaryTooltip term="session">Sessions API</GlossaryTooltip> for read replication, all of your queries from a single Session read from a version of the database which is as up-to-date as your query. This ensures that the version of the database you are reading is logically consistent with your queries when using read replicas.

### Bookmarks

> D1 read replication achieves the total ordering of all operations by attaching a bookmark to each write query within a Session. A bookmark represents the state of a database at a specific point in time. For more information, refer to [Bookmarks](/d1/reference/time-travel/#bookmarks).

Author: - Added `bookmark` to glossary, with definition: `A bookmark represents the state of a database at a specific point in time.`
- In the Time Travel chapter, it also says the bookmark `is effectively an append-only log.`. If I remember correctly @lambrospetrou this needed rewording?

Subsequent interactions on the application should continue using the same Session by passing the `bookmark` from the first query to subsequent requests. This guarantees that all interactions will observe information at least as up-to-date as the initial request, and therefore never show information older than what a user has already observed. The number of likes will be updated with newer counts over time with subsequent requests as D1 asynchronously updates the read replicas with the changes from the primary database.

```js
async function getLikes(postId: string, db: D1Database, bookmark: string | null): GetLikesResult {
  // NOTE: Achieve sequential consistency with given bookmark,
  //       or start a new session that can be served by any replica.
  const session = db.withSession(bookmark ?? "first-unconstrained");
  const { results } = session.prepare("SELECT * FROM likes WHERE postId = ?").bind(postId).run();
  return { bookmark: session.getBookmark(), likes: results };
}
```

## Read replica locations

> Currently, D1 automatically creates a read replica in every region, including the region where the primary database instance is located. These regions are:

Reviewer: link to https://developers.cloudflare.com/d1/configuration/data-location/#available-location-hints to make connection obvious

- ENAM
- WNAM
- WEUR
- EEUR
- APAC
- OC

> ## Benefits of read replication

Reviewer: Could group this alongside `Replica lag` section

A system with multiple read replicas located around the world improves the performance of databases:

- The read throughput increases by distributing load across multiple replicas. Since multiple database instances are able to serve read-only requests, your application can serve a larger number of queries at any given time.
- The query latency decreases for users located close to the read replicas. By shortening the physical distance between a read replica and the user, read query latency decreases, resulting in a faster application.

> ## Monitor latency

Reviewer: Rename `Observability`, should mention:

- new query results `meta` fields (servedByRegion, servedByPrimary, bookmark)
- UI metrics breakdown by region

To see the impact of read replication, check the latency for your read-only queries. Refer to the [queryBatchTimeMs](/d1/observability/metrics-analytics/) metric or [view your metrics through the dashboard](/d1/observability/metrics-analytics/#view-metrics-in-the-dashboard).

## Known limitations

There are some known limitations for D1 read replication.

- The lifecycle of a read replica is tied to the lifecycle of the primary database instance. For example, iof the primary database instance is inactive, the read replicas are inactive too.
- TBC

## Supplementary information

You may wish to refer to the following resources:

- Blog: [Building D1: a Global Database](https://blog.cloudflare.com/building-d1-a-global-database/)
- [D1 Sessions API documentation](/d1/worker-api/d1-database#withsession)

> - [Read replication demo application](/d1/demos/)

Reviewer: demos to be included are: https://chat.google.com/room/AAAASFG24Rc/BAUz4LlkhnA/BAUz4LlkhnA?cls=10

---
title: Read replication
pcx_content_type: concept
sidebar:
  order: 90

---

import { GlossaryTooltip, Details, GitHubCode, APIRequest } from "~/components"

D1 read replication scales read throughput and lowers latency for read queries by adding read-only replicas across regions closer to clients.

You can use D1 read replication by using D1 Sessions API. A Session encapsulates all the queries from one logical session for your application. For example, a Session may correspond to all queries coming from a particular web browser session.

> By using Sessions API for read replication, all of your queries from a single Session read from a version of the database which is as up-to-date as your query requires. This ensures that the version of the database you are reading is logically consistent with your queries when using read replicas.

Reviewer 2: This entire line is confusing.

The following Worker code uses D1 Sessions API to use read replicas, and allows you to test the decrease in query latency when using read replication.

```ts collapse={1-6, 34-130}
import { D1Database, D1DatabaseSession } from "@cloudflare/workers-types";

export type Env = {
	DB01: D1Database;
};

export default {
	async fetch(request, env, ctx): Promise<Response> {
		const url = new URL(request.url);

		// A. Create the Session.
		// When we create a D1 Session, we can continue where we left off from a previous
		// Session if we have that Session's last bookmark or use a constraint.
		const bookmark = request.headers.get('x-d1-bookmark') ?? 'first-unconstrained';
		const session = env.DB01.withSession(bookmark);

		try {
			// Use this Session for all our Workers' routes.

> 			const response = await withTablesInitialized(session, async () => await handleRequest(request, session));

Reviewer 2: This is going to be in the main doc for read replication. The `withTablesInitialized` I have in https://github.com/lambrospetrou/d1-starter-sessions-api/blob/114186685734ef8b1ae5e581c731f2ee214c706f/src/index.ts#L19 is mostly to make the starter demo easier to deploy without having to run migrations.

All production projects will probably not do this, so you might not want this to be in the snippet shown in docs here.

Author: I see, Is there a different snippet we can use?

Reviewer 2: It depends what we want to highlight.

Showing 4 lines of sessions api is not the same as production code where you have error handling, and migrations to run and all that.

If everything is to be using the Deploy to Workers button then either there should be a first step to run migrations or the code itself will have to do it as I do in the demo.

Not every snippet of code in the docs has to be a fully deployable project in my opinion.

			await initTables(session);
			return await handler();
		}
		throw e;
	}
}

async function initTables(session: D1DatabaseSession) {
	return await session
		.prepare(`CREATE TABLE IF NOT EXISTS Orders(
			customerId TEXT NOT NULL,
			orderId TEXT NOT NULL,
			quantity INTEGER NOT NULL,
			PRIMARY KEY (customerId, orderId)
		)`)
		.all();
}

async function resetTables(session: D1DatabaseSession) {
	return await session
		.prepare(`DROP TABLE IF EXISTS Orders; CREATE TABLE IF NOT EXISTS Orders(
			customerId TEXT NOT NULL,
			orderId TEXT NOT NULL,
			quantity INTEGER NOT NULL,
			PRIMARY KEY (customerId, orderId)
		)`)
		.all();
}
```

{/* Deploy to Workers button */}

## Primary database instance vs read replicas

When using D1 without read replication, D1 routes all queries (both read and write) to a specific database instance in [one location in the world](/d1/configuration/data-location/), known as the primary database instance. The request latency is dependent on the physical closeness of a user to the primary database instance. Users located further away from this database instance experience the longest request latency.

> When using read replication, D1 introduces multiple “almost up-to-date” copies of the primary database instance which only serve read requests, called <GlossaryTooltip term="read replica"> read replicas </GlossaryTooltip>. D1 creates the read replicas in multiple regions throughout the world [across the Cloudflare network](/d1/features/read-replication/#read-replica-locations).

Reviewer 2: Let's stop using the term "almost up-to-date", please. It doesn't give any technical understanding of what it means. And if there isn't a stream of constant writes in the database, it's not "almost up-to-date", it's fully up-to-date.

We should use the proper term which is "replicated copies" or "eventually replicated copies", or similar.

> When using read replication, D1 introduces multiple “almost up-to-date” copies of the primary database instance which only serve read requests, called <GlossaryTooltip term="read replica"> read replicas </GlossaryTooltip>. D1 creates the read replicas in multiple regions throughout the world [across the Cloudflare network](/d1/features/read-replication/#read-replica-locations).

Author: @vy-ton do you have a preference on which of the suggested terms you want to use going forward?

A user may be located far away from the primary database instance but close to a read replica. When D1 routes read requests to the read replica instead of the primary database instance, the user enjoys shorter read request response times.

> D1 asynchronously updates read replicas against the primary database instance. This means that at any given time, a read replica may be arbitrarily out of date. The difference between the primary database instance and the read replica is known as the <GlossaryTooltip term="replica lag"> replica lag </GlossaryTooltip>. For more information, refer to [Replica lag and consistency model](/d1/features/read-replication/#replica-lag-and-consistency-model).

Reviewer 2: ```suggestion
D1 asynchronously replicates changes from the primary database instance to all read replicas. This means that at any given time, a read replica may be out of date. The time it takes for the latest change in the primary database to be replicated to a read replica is known as the <GlossaryTooltip term="replica lag"> replica lag </GlossaryTooltip>. For more information, refer to [Replica lag and consistency model](/d1/features/read-replication/#replica-lag-and-consistency-model).
```

![D1 read replication concept](/images/d1/d1-read-replication-concept.png)

:::note
All write queries are still forwarded to the primary database instance. Read replication only improves the query response time for read requests.
:::

| Type of database instance | Description                                                                                                                                       | How it handles write queries                                | How it handles read queries                               |
| ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------- | --------------------------------------------------------- |
| Primary database instance | The database instance containing the “original” copy of the database                                                                              | Can serve write queries                                     | Can serve read queries                                    |

> | Read replica              | A database instance containing an “almost up-to-date” copy of the original database, asynchronously updated against the primary database instance | Forwards any write queries to the primary database instance | Can serve read queries using its own copy of the database |

Reviewer 2: See my suggestions in the glossary

### Check if read replication is enabled

The following code uses REST API to check if read replication is enabled / disabled.

```ts
const headers = new Headers({
  "Authorization": `Bearer ${TOKEN}`
});

const response = await fetch("/v4/accounts/{account_id}/d1/database/{database_id}", {
  method: "GET",
  headers: headers
});

const data = await response.json();
console.log(data.read_replication.mode);
```

- If the output is `auto`, read replication is enabled.
- If the output is `disabled`, read replication is disabled.

### Enable or disable read replication

Use REST API to enable / disable read replication on your D1 database.

```ts
const headers = new Headers({
  "Authorization": `Bearer ${TOKEN}`
});

await fetch ("/v4/accounts/{account_id}/d1/database/{database_id}", {
	method: "PUT",
	headers: headers,
	body: JSON.stringify(
		{ "read_replication": { "mode": "auto" } }
	)
 }
)
```

- Set `"mode":"auto"` to enable read replication.
- Set `"mode":"disable"` to disable read replication.

If this is your first time using REST API, create an API token. Refer to [Create API token](/fundamentals/api/get-started/create-token/).

> ### Start a D1 Session without constraints

Reviewer: everything in examples can just say `Session` not `D1 Session`

> You can start a new D1 Session without constraints. D1 will route the first read query in the Session to any database instance, which could be either the primary database instance or a read replica.

Reviewer: ```suggestion
To create a Session from any database version, use `withSession()` without any parameters, which will route the first query to any database instance, either the primary database instance or a read replica. 
```
Everything before the code snippet can be said with this, aim for 1-2 sentences max

Use this option if you do not need to read the latest version of the database in the primary database instance.

To do this, do not pass any parameters to `withSession()`.

```ts
// synchronous
let session = env.DB_D1.withSession()

> const stmt = session

Reviewer: Refer to Lambros' example for correction code

```session.prepare.run()```

.prepare(`SELECT * FROM Customers WHERE CompanyName = 'Bs Beverages'`)
// wait for Session condition, queries within batch have casual consistency
const result = await session.run()
```

> - D1 serves the first read query may immediately start using read replicas. This will be a version of the database that is as up-to-date as the query needs it to be.

Reviewer: As is these bullets confuse or is too repetitive

What a user needs to know:
```
* `withSession()` same as `withSession("first-unconstrained")`
* Best for when your application does not require the latest database version. All queries in a Session ensure sequential consistency
* Refer to API docs
```

- Possible use-case: Displaying blog articles on a website. It is not crucial to display the latest blog article, and therefore the user can start the Session by reading from a read replica. Subsequent queries can also go to the read replica.
- Refer to the [D1 Workers Binding API documentation](/d1/worker-api/d1-database#withsession).

> #### Example of a D1 Session without constraints

Reviewer: Suggestion drop these out for now, and we can bring in after the 1-pager foundation is solid

Subsequent interactions on the application should continue using the same Session by passing the `bookmark` from the first query to subsequent requests. This guarantees that all interactions will observe information at least as up-to-date as the initial request, and therefore never show information older than what a user has already observed. The number of likes will be updated with newer counts over time with subsequent requests as D1 asynchronously updates the read replicas with the changes from the primary database.

```js
async function getLikes(postId: string, db: D1Database, bookmark: string | null): GetLikesResult {
  // NOTE: Achieve sequential consistency with given bookmark,
  //       or start a new session that can be served by any replica.
  const session = db.withSession(bookmark ?? "first-unconstrained");
  const { results } = session
	.prepare("SELECT * FROM likes WHERE postId = ?")
	.bind(postId)
	.run();
  return { bookmark: session.getBookmark(), likes: results };
}
```

> ### Start a D1 Session with a constraint

Reviewer: ```suggestion
### Start a Session with all latest data
```

> You can start a new D1 Session by forcing the first read query to go to the primary database instance. Subsequent read queries may read from the read replica.

Reviewer: Use my example for `Start a Session without constraints` to edit this one

```ts
async function listBillStatements(accountId: string, db: D1Database): Promise<ListBillStatementsResult> {
	const session = db.withSession('first-primary');
	const { results } = (await session.prepare('SELECT * FROM bills WHERE accountId = ?').bind(accountId).run()) as unknown as {
		results: Bill[];
	};
	return { bookmark: session.getBookmark() ?? 'first-unconstrained', bills: results };
}

async function getBillStatement(accountId: string, billId: string, bookmark: string, db: D1Database): Promise<GetBillStatementResult> {
	// NOTE: We achieve sequential consistency with the given `bookmark`.
	const session = db.withSession(bookmark);
	const result = (await session
		.prepare('SELECT * FROM bills WHERE accountId = ? AND billId = ? LIMIT 1')
		.bind(accountId, billId)
		.first()) as unknown as Bill;

	return { bookmark: session.getBookmark() ?? 'first-unconstrained', bill: result };
}
```

> ### Continue a D1 Session with a bookmark

Reviewer: helps users understand a Session can't be restarted

> You can continue an existing Session from a previous Session.

Reviewer: ```suggestion
To create a new Session from the context of a previous Session, pass a `bookmark` parameter to guarantee that the Session starts with database state as of or later than the provided `bookmark`.
```

Use this option if you need to read from the specific version of the database which you last read from your previous Session.

To do this, pass the `bookmark` string to `withSession()`.

```ts
// synchronous
let session = env.DB_D1.withSession('bookmark_string')
const stmt = session
.prepare(`SELECT * FROM Customers WHERE CompanyName = 'Bs Beverages'`)
// wait for Session condition, queries within batch have casual consistency
const result = await session.run()
```

> - Starting a Session with a `bookmark` continues an existing Session using the `bookmark` as the reference point. This ensures you read from a version of the database which is at least as up-to-date as the provided `bookmark`.

Reviewer: It doesn't continue an existing Session. `bookmark` ensures it has all the same context as the last Session

- Possible use-case: Reading the score-table of an ongoing sports tournament (a continuation of the example from [Start a D1 Session with a constraint](/d1/features/read-replication/#start-a-d1-session-with-a-constraint)). A user may have started a web browser session and already has a `bookmark`. D1 can serve subsequent read queries within the same Session which is logically consistent, by using the `bookmark`.
- Refer to the [D1 Workers Binding API documentation](/d1/worker-api/d1-database#withsession).

> #### Example of using `bookmark`

Reviewer: This needs to be the code snippet for HTTP cookie but just the parts of:
- get bookmark
- store bookmark in cookier
- start a session from bookmark

Reviewer: code snippets for tasks should not be full, end-to-end examples

Reviewer 2: I personally don't think we need to involve cookie code in this doc. Each framework treats cookies in its own way, and there are many NPM libraries to manage cookies. Once we show how to do it with HTTP headers or request/response bodies folks should be able to port it in whatever cookie library they use.

	const session = db.withSession(bookmark);
	const result = (await session
		.prepare('SELECT * FROM bills WHERE accountId = ? AND billId = ? LIMIT 1')
		.bind(accountId, billId)
		.first()) as unknown as Bill;

	return { bookmark: session.getBookmark() ?? 'first-unconstrained', bill: result };
}

export default {
	async fetch(request, env, ctx): Promise<Response> {
		// URL path
		const url = new URL(request.url);
		const path = url.pathname;

		// Method
		const method = request.method;

		// Fetch using first-unconstrained
		if (path === '/bills' && method === 'GET') {
			// List bills
			const result = await listBillStatements('1', env.DB);
			return new Response(JSON.stringify(result), { status: 200 });
		}
		if (path === '/bill' && method === 'GET') {
			// Get bill
			const result = await getBillStatement('1', '1', 'first-unconstrained', env.DB);
			return new Response(JSON.stringify(result), { status: 200 });
		}

		// Fetch using bookmark from cookie
		if (path === '/bill/cookie' && method === 'GET') {
			// Get bill
			const cookie = request.headers.get('Cookie');
			const bookmark =
				cookie
					?.split(';')
					.find((c) => c.trim().startsWith('X-D1-Bookmark'))

> 					?.split('=')[1] ?? 'first-unconstrained';

Reviewer 2: IF you are going to keep the cookies example, please use a proper library that uses cookies correctly, and not doing cookie parsing manually.

Author: CC @vy-ton, @harshil1712

			console.log('bookmark', bookmark);
			const result = await getBillStatement('1', '1', bookmark, env.DB);
			return new Response(JSON.stringify(result), {
				status: 200,
				headers: {
					'Set-Cookie': `X-D1-Bookmark=${result.bookmark}; Path=/; SameSite=Strict`,
				},
			});
		}

		// To ingest data
		if (path === '/bill' && method === 'POST') {
			// Create bill
			const { accountId, amount, description, due_date } = await request.json();
			const session = env.DB.withSession('first-primary');
			const { results } = await session
				.prepare('INSERT INTO bills (accountId, amount, description, due_date) VALUES (?, ?, ?, ?) RETURNING *')
				.bind(accountId, amount, description, due_date)
				.run();
			const bookmark = session.getBookmark() ?? 'first-unconstrained';

			return new Response(JSON.stringify(results), {
				status: 201,
				headers: {
					// Set bookmark cookie
					'Set-Cookie': `X-D1-Bookmark=${bookmark}; Path=/; SameSite=Strict`,
				},
			});
		}
		return new Response('Not Found', {
			status: 404,
			statusText: 'Not Found',
		});
	},
} satisfies ExportedHandler<Env>;
```

## Read replica locations

> 

Reviewer: Add note that replica locations are subject to be changed by Cloudflare

Currently, D1 automatically creates a read replica in [every supported region](/d1/configuration/data-location/#available-location-hints), including the region where the primary database instance is located. These regions are:

- ENAM
- WNAM
- WEUR
- EEUR
- APAC
- OC

## Observability

To see the impact of read replication, check the latency for your read-only queries. You can use:

> - The `meta` object within the [`D1Result`](/d1/worker-api/return-object/#d1result) return object, which includes new fields:

Reviewer: lets also add a example/task of introspecting meta fields, can use code from https://github.com/lambrospetrou/d1-starter-sessions-api/blob/main/src/index.ts#L80

>   - `servedByRegion`

Reviewer 2: ```suggestion
  - `served_by_region`
```

>   - `servedByPrimary`

Reviewer 2: ```suggestion
  - `served_by_primary`
```

>   - `bookmark`

Reviewer: bookmark not on `meta`

- The [Cloudflare dashboard](https://dash.cloudflare.com/), where you can view your metrics breakdown by region.

## Known limitations

There are some known limitations for D1 read replication.

> - The lifecycle of a read replica is tied to the lifecycle of the primary database instance. For example, iof the primary database instance is inactive, the read replicas are inactive too.

Reviewer 2: I am certain this will confuse people. Do people care for this? How does this affect their code or how they use read replication?

It's an implementation detail on our side.

- TBC

## Background information

### Replica lag and consistency model

To account for replica lag, it is important to consider the consistency model for D1. A consistency model is a logical framework that governs how a database system serves user queries (how the data is updated and accessed) when there are multiple database instances. Different models can be useful in different use cases. Most database systems provide [read committed](https://jepsen.io/consistency/models/read-committed), [snapshot isolation](https://jepsen.io/consistency/models/snapshot-isolation), or [serializable](https://jepsen.io/consistency/models/serializable) consistency models, depending on their configuration.

> D1 read replication offers [sequential consistency](https://jepsen.io/consistency/models/sequential). D1 creates a global order of all operations which have taken place on the database, and can identify the latest version of the database that a query has seen, using [bookmarks](/d1/reference/time-travel/#bookmarks). It then serves the query with a database instance that is at least as up-to-date as the query itself.

Reviewer 2: ```suggestion
D1 read replication offers [sequential consistency](https://jepsen.io/consistency/models/sequential). D1 creates a global order of all operations which have taken place on the database, and can identify the latest version of the database that a query has seen, using [bookmarks](/d1/reference/time-travel/#bookmarks). It then serves the query with a database instance that is at least as up-to-date as the bookmark requested.
```

Sequential consistency has properties such as:

- **Monotonic reads**: If you perform two reads one after the other (read-1, then read-2), read-2 cannot read a version of the database prior to read-1.
- **Monotonic writes**: If you perform write-1 then write-2, all processes observe write-1 before write-2.
- **Writes follow reads**: If you read a value, then perform a write, the subsequent write must be based on the value that was just read.
- **Read my own writes**: If you write to the database, all subsequent reads will see the write.

### Benefits of read replication

A system with multiple read replicas located around the world improves the performance of databases:

- The read throughput increases by distributing load across multiple replicas. Since multiple database instances are able to serve read-only requests, your application can serve a larger number of queries at any given time.

> - The query latency decreases for users located close to the read replicas. By shortening the physical distance between a read replica and the user, read query latency decreases, resulting in a faster application.

Reviewer 2: ```suggestion
- The query latency decreases for users located close to the read replicas. By shortening the physical distance between the database and the user, read query latency decreases, resulting in a faster application.
```

## Supplementary information

You may wish to refer to the following resources:

> - Blog: [Building D1: a Global Database](https://blog.cloudflare.com/building-d1-a-global-database/)

Reviewer 2: And the new blog post obviously.
